Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCell.cs

```diff
@@ -0,0 +1,1265 @@
+            Contract.Requires(rowIndex >= 0 || rowIndex < this.DataGridView.Rows.Count);
+            Contract.Requires(this.ColumnIndex < 0);
+            return default(ContextMenuStrip);
+        }
+
+        /// <summary>
+        /// Returns a value indicating the current state of the cell as inherited from the state of its row and column.
+        /// </summary>
+        /// 
+        /// <returns>
+        /// A bitwise combination of <see cref="T:System.Windows.Forms.DataGridViewElementStates"/> values representing the current state of the cell.
+        /// </returns>
+        /// <param name="rowIndex">The index of the row containing the cell.</param><exception cref="T:System.ArgumentException">The cell is not contained within a <see cref="T:System.Windows.Forms.DataGridView"/> control and <paramref name="rowIndex"/> is not -1.-or-<paramref name="rowIndex"/> is not the index of the row containing this cell.</exception><exception cref="T:System.ArgumentOutOfRangeException">The cell is contained within a <see cref="T:System.Windows.Forms.DataGridView"/> control and <paramref name="rowIndex"/> is outside the valid range of 0 to the number of rows in the control minus 1.</exception>
+        public virtual DataGridViewElementStates GetInheritedState(int rowIndex)
+        {
+            Contract.Requires(this.DataGridView != null);
+            Contract.Requires(rowIndex >= 0 || rowIndex < this.DataGridView.Rows.Count);
+            return default(DataGridViewElementStates);
+        }
+
+        /// <summary>
+        /// Gets the style applied to the cell.
+        /// </summary>
+        /// 
+        /// <returns>
+        /// A <see cref="T:System.Windows.Forms.DataGridViewCellStyle"/> that includes the style settings of the cell inherited from the cell's parent row, column, and <see cref="T:System.Windows.Forms.DataGridView"/>.
+        /// </returns>
+        /// <param name="inheritedCellStyle">A <see cref="T:System.Windows.Forms.DataGridViewCellStyle"/> to be populated with the inherited cell style. </param><param name="rowIndex">The index of the cell's parent row. </param><param name="includeColors">true to include inherited colors in the returned cell style; otherwise, false. </param><exception cref="T:System.InvalidOperationException">The cell has no associated <see cref="T:System.Windows.Forms.DataGridView"/>.-or-<see cref="P:System.Windows.Forms.DataGridViewCell.ColumnIndex"/> is less than 0, indicating that the cell is a row header cell.</exception><exception cref="T:System.ArgumentOutOfRangeException"><paramref name="rowIndex"/> is less than 0, or greater than or equal to the number of rows in the parent <see cref="T:System.Windows.Forms.DataGridView"/>.</exception>
+        public virtual DataGridViewCellStyle GetInheritedStyle(DataGridViewCellStyle inheritedCellStyle, int rowIndex,
+            bool includeColors)
+        {
+            Contract.Requires(this.DataGridView != null);
```

Comment:
This is the only method for now that trully have this precondition! (see http://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/DataGridViewCell.cs,48f5a95ad5483bc1, this method throws InvalidOperationException when DataGridView is null).